content/blog/kubecon-2019/index.md

```diff
@@ -0,0 +1,272 @@
+                        { name: "PROJECT", value: project },
+                        {
+                            name: "GOOGLE_APPLICATION_CREDENTIALS",
+                            value: "/var/secrets/google/key.json"
+                        },
+                    ],
+                    ...
+                }]
+            }
+        }
+    },
+});
+```
+
+### Acceptance Tests
+
+Our acceptance tests are what validate that our service when stood up function as expected. They are run against an ephemeral environment. The test live in the `acceptance/acceptance_test.go` [file](https://gitlab.com/rocore/demo-app/blob/master/acceptance/acceptance_test.go). You'll notice we're once again using that helper function `getConfigurationValue`. Our acceptance test must also be configured to ensure they're validating against the correct resources for that particular ephemeral environment.
+
+Since our service is only accessible from within the Kubernetes cluster, we will use a Kubernetes Job to run our acceptance tests. This is a good technique to use when your CI is running externally, such as from GitLab, and you do not want to publicly expose your service. Our ephemeral environment plus acceptance test look like this:
+
+![acceptance test architecture](acceptance-test-arch.jpg)
+
+We spin up our Kubernetes Job and additional resources by using an if statement at the bottom of our `infrastructure/index.ts` file. The conditional depends on the environment's name as follows:
+
+```typescript
+// If it's a test environment, set up acceptance tests.
+let job: k8s.batch.v1.Job | undefined;
+if (ENV.startsWith("test")) {
+    job = acceptance.setupAcceptanceTests({
+        ...
+    });
+}
+
+// Export the acceptance job name, so we can get the logs from our
+// acceptance tests.
+export const acceptanceJobName = job ? job.metadata.name : "unapplicable";
+```
+
+That covers all the major aspects of our application and infrastructure, and if you'd like to view the code in detail it is all available for reference in our `demo-app` [GitLab repository](https://gitlab.com/rocore/demo-app).
+
+## Our Pipeline
+
+When developing a new service, it’s important that we establish a solid deployment strategy upfront. We want to make sure we're building in quality from day one. As we develop the service, we can add acceptance tests for each and every feature we add while the context and requirements are still fresh in our minds. This ensures we have thorough coverage of our app's functionality.
+
+We used GitLab to set up our pipeline. We chose GitLab because its extremely easy to set up and allows us to run our pipeline on our Docker image of choice. We use a [base-image](https://gitlab.com/rocore/global-infra/blob/master/base-image/Dockerfile) that has all our dependencies installed and then reference that Docker image and tag in our `demo-app` pipeline. This allows us to bundle and version the dependencies for building our application and infrastructure.
```

Comment:
```suggestion
We used GitLab to set up our pipeline. We chose GitLab because it's extremely easy to set up and allows us to run our pipeline on our Docker image of choice. We use a [base-image](https://gitlab.com/rocore/global-infra/blob/master/base-image/Dockerfile) that has all our dependencies installed and then reference that Docker image and tag in our `demo-app` pipeline. This allows us to bundle and version the dependencies for building our application and infrastructure.
```